test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/AuthenticationHandlerTests.cs

```diff
@@ -0,0 +1,215 @@
+                        _ = await source.GetAsync(request, ProcessResponse, logger.Object, cancellationToken);
+                    }
+                }
+                finally
+                {
+                    server.Stop();
+                }
+
+                // Assert
+                Assert.Equal(
+                    1,
+                    server.Requests.Count(RequestWithoutAuthorizationHeader));
+                Assert.Equal(
+                    2,
+                    server.Requests.Select(r => r.RawUrl).Distinct().Count());
+
+                // ExecuteAsync returns Task<T>, so need to return something to give it a <T>.
+                return (object?)null;
+            },
+            CancellationToken.None);
+
+            static bool RequestWithoutAuthorizationHeader(HttpListenerRequest request)
+            {
+                string? value = request.Headers["Authorization"];
+                return string.IsNullOrEmpty(value);
+            }
+
+            static Task<string> ProcessResponse(HttpSourceResult result)
+            {
+                return Task.FromResult(string.Empty);
+            }
+        }
+
+        private class RequestCollectingServer
+        {
+            private string? _baseUrl;
+            private HttpListener? _httpListener;
+            private List<HttpListenerRequest> _requests = new();
+            private ITestOutputHelper _output;
+
+            public RequestCollectingServer(ITestOutputHelper output)
+            {
+                _output = output;
+            }
+
+            public string BaseUrl
+            {
+                get
+                {
+                    if (_baseUrl == null)
+                    {
+                        throw new InvalidOperationException("Start must be called first");
+                    }
+                    return _baseUrl;
+                }
+            }
+
+            public IReadOnlyList<HttpListenerRequest> Requests
+            {
+                get
+                {
+                    return _requests;
+                }
+            }
+
+            public void Start(int port)
+            {
+                if (_httpListener != null)
+                {
+                    throw new InvalidOperationException("Already started");
+                }
+
+                string baseUrl = $"http://localhost:{port}/";
+                var listener = new HttpListener();
+                listener.Prefixes.Add(baseUrl);
+                listener.AuthenticationSchemes = AuthenticationSchemes.Basic | AuthenticationSchemes.Anonymous;
+                listener.Start();
+                _httpListener = listener;
+                _baseUrl = baseUrl;
+
+                listener.BeginGetContext(EndGetContext, listener);
+            }
+
+            public void EndGetContext(IAsyncResult result)
```

Comment:
nit: Using a name like `ProcessRequest` for this method would be a better option, as it can help readers to easily understand that this method is responsible for processing the http request. Not a blocking comment.

Reply:
I agree because HttpListener uses this name for a function that returns a context, whereas this is just a class method to execute.